File: internal/shmem/channel.go
// Copyright 2026 The go-hbase Authors
// SPDX-License-Identifier: Apache-2.0

package shmem

import (
	"errors"
	"fmt"

	"github.com/viroge/go-shmem/pkg/config"
	"github.com/viroge/go-shmem/pkg/memory"
	"github.com/viroge/go-shmem/pkg/message"
	"github.com/viroge/go-shmem/pkg/ring"
	"github.com/viroge/go-shmem/pkg/ring/consumer"
	"github.com/viroge/go-shmem/pkg/ring/producer"
)

type Role uint8

const (
	RoleProducer Role = iota + 1
	RoleConsumer
)

type Frame = []byte

type Config struct {
	Backend       string
	Filename      string
	ShmName       string
	Capacity      int
	MaxObjectSize int
	Role          Role
}

type Channel struct {
	role           Role
	maxPayloadSize int
	msg            *message.PayloadMessage

	prod *producer.Producer
	cons *consumer.Consumer
}

var (
	ErrRingFull      = ring.ErrRingFull
	ErrNoData        = ring.ErrNoData
	ErrFrameTooLarge = errors.New("shmem: frame exceeds max payload size")
	ErrWrongRole     = errors.New("shmem: operation not permitted for this role")
)

func Open(cfg Config) (*Channel, error) {
	r, err := toRingConfig(cfg)
	if err != nil {
		return nil, err
	}

	maxPayload := cfg.MaxObjectSize - 4 // 4-byte length prefix per PayloadMessage layout
	if maxPayload <= 0 {
		return nil, fmt.Errorf("shmem: MaxObjectSize=%d leaves no room for payload", cfg.MaxObjectSize)
	}

	ch := &Channel{
		role:           cfg.Role,
		maxPayloadSize: maxPayload,
		msg:            message.NewPayloadMessage(maxPayload),
	}

	switch cfg.Role {
	case RoleProducer:
		p, err := producer.Open(r)
		if err != nil {
			return nil, fmt.Errorf("shmem: open producer: %w", err)
		}
		ch.prod = p
	case RoleConsumer:
		c, err := consumer.Open(r)
		if err != nil {
			return nil, fmt.Errorf("shmem: open consumer: %w", err)
		}
		ch.cons = c
	default:
		return nil, fmt.Errorf("shmem: invalid role: %d", cfg.Role)
	}

	return ch, nil
}

func (c *Channel) Send(f Frame) error {
	if c.role != RoleProducer {
		return ErrWrongRole
	}
	if len(f) > c.maxPayloadSize {
		return fmt.Errorf("%w: %d > %d", ErrFrameTooLarge, len(f), c.maxPayloadSize)
	}

	addr, err := c.prod.NextToDispatch()
	if err != nil {
		return err
	}

	if err := c.msg.SetPayload(f); err != nil {
		return err
	}
	c.msg.WriteTo(addr)
	c.prod.Flush()
	return nil
}

func (c *Channel) Recv() (Frame, error) {
	if c.role != RoleConsumer {
		return nil, ErrWrongRole
	}

	addr, err := c.cons.FetchNext()
	if err != nil {
		return nil, err
	}

	if err := c.msg.ReadFrom(addr); err != nil {
		c.cons.RollBack()
		return nil, err
	}

	payload := c.msg.GetPayload()
	out := make([]byte, len(payload))
	copy(out, payload)

	c.cons.DoneFetching()
	return out, nil
}

func (c *Channel) Close() error {
	switch {
	case c.prod != nil:
		err := c.prod.Close(false)
		c.prod = nil
		return err
	case c.cons != nil:
		err := c.cons.Close(false)
		c.cons = nil
		return err
	}
	return nil
}

func toRingConfig(cfg Config) (config.Ring, error) {
	if cfg.Capacity <= 0 {
		return config.Ring{}, fmt.Errorf("shmem: Capacity must be > 0, got %d", cfg.Capacity)
	}
	if cfg.MaxObjectSize <= 0 {
		return config.Ring{}, fmt.Errorf("shmem: MaxObjectSize must be > 0, got %d", cfg.MaxObjectSize)
	}

	backend := cfg.Backend
	if backend == "" {
		backend = memory.BackendMmap
	}

	switch backend {
	case memory.BackendMmap:
		if cfg.Filename == "" {
			return config.Ring{}, errors.New("shmem: Filename required for mmap backend")
		}
	case memory.BackendPosixShm:
		if cfg.ShmName == "" {
			return config.Ring{}, errors.New("shmem: ShmName required for posix_shm backend")
		}
	default:
		return config.Ring{}, fmt.Errorf("shmem: unknown backend %q", backend)
	}

	return config.Ring{
		Backend:       backend,
		Filename:      cfg.Filename,
		ShmName:       cfg.ShmName,
		Capacity:      cfg.Capacity,
		MaxObjectSize: cfg.MaxObjectSize,
	}, nil
}

File: java/com/virogg/hbasecop/bridge/shmem/Channel.java
// Copyright 2026 The go-hbase Authors
// SPDX-License-Identifier: Apache-2.0

package com.virogg.hbasecop.bridge.shmem;

import com.jgshmem.message.PayloadMessage;
import com.jgshmem.ring.WaitingRingConsumer;
import com.jgshmem.ring.WaitingRingProducer;
import com.jgshmem.utils.Builder;
import java.util.Objects;
import java.util.Optional;

public final class Channel implements AutoCloseable {

  private final Role role;
  private final int maxPayloadSize;
  private final WaitingRingProducer<PayloadMessage> producer;
  private final WaitingRingConsumer<PayloadMessage> consumer;

  private Channel(
      Role role,
      int maxPayloadSize,
      WaitingRingProducer<PayloadMessage> producer,
      WaitingRingConsumer<PayloadMessage> consumer) {
    this.role = role;
    this.maxPayloadSize = maxPayloadSize;
    this.producer = producer;
    this.consumer = consumer;
  }

  public static Channel open(Config cfg) {
    validate(cfg);

    int maxPayload = cfg.maxObjectSize() - 4; // 4-byte length prefix per PayloadMessage layout
    if (maxPayload <= 0) {
      throw new IllegalArgumentException(
          "shmem: maxObjectSize=" + cfg.maxObjectSize() + " leaves no room for payload");
    }

    com.jgshmem.config.Config.Ring ring = toUpstreamRing(cfg);
    Builder<PayloadMessage> msgBuilder = () -> new PayloadMessage(maxPayload);

    if (cfg.role() == Role.PRODUCER) {
      WaitingRingProducer<PayloadMessage> p =
          new WaitingRingProducer<>(cfg.capacity(), cfg.maxObjectSize(), msgBuilder, ring);
      return new Channel(Role.PRODUCER, maxPayload, p, null);
    }

    WaitingRingConsumer<PayloadMessage> c =
        new WaitingRingConsumer<>(cfg.capacity(), cfg.maxObjectSize(), msgBuilder, ring);
    return new Channel(Role.CONSUMER, maxPayload, null, c);
  }

  public int capacity() {
    return role == Role.PRODUCER ? producer.getCapacity() : consumer.getCapacity();
  }

  public void send(byte[] frame) throws ShmemException {
    if (role != Role.PRODUCER) {
      throw new IllegalStateException("shmem: send on " + role + " channel");
    }
    Objects.requireNonNull(frame, "frame");
    if (frame.length > maxPayloadSize) {
      throw new FrameTooLargeException(frame.length, maxPayloadSize);
    }

    PayloadMessage msg = producer.nextToDispatch();
    if (msg == null) {
      throw new RingFullException();
    }
    msg.payloadSize = frame.length;
    msg.payload.clear();
    if (frame.length > 0) {
      msg.payload.put(frame);
    }
    msg.payload.flip();
    producer.flush();
  }

  public Optional<byte[]> recv() {
    if (role != Role.CONSUMER) {
      throw new IllegalStateException("shmem: recv on " + role + " channel");
    }
    if (consumer.availableToFetch() <= 0) {
      return Optional.empty();
    }

    PayloadMessage msg = consumer.fetch();
    byte[] out = new byte[msg.payloadSize];
    if (msg.payloadSize > 0) {
      msg.payload.position(0);
      msg.payload.limit(msg.payloadSize);
      msg.payload.get(out);
    }
    consumer.doneFetching();
    return Optional.of(out);
  }

  @Override
  public void close() {
    if (producer != null) {
      producer.close(false);
    } else if (consumer != null) {
      consumer.close(false);
    }
  }

  private static void validate(Config cfg) {
    Objects.requireNonNull(cfg, "cfg");
    if (cfg.role() == null) {
      throw new IllegalArgumentException("shmem: role required");
    }
    if (cfg.capacity() <= 0) {
      throw new IllegalArgumentException("shmem: capacity must be > 0, got " + cfg.capacity());
    }
    if (cfg.maxObjectSize() <= 0) {
      throw new IllegalArgumentException(
          "shmem: maxObjectSize must be > 0, got " + cfg.maxObjectSize());
    }

    String backend = cfg.backend() == null ? Config.BACKEND_MMAP : cfg.backend();
    switch (backend) {
      case Config.BACKEND_MMAP:
        if (cfg.filename() == null || cfg.filename().isEmpty()) {
          throw new IllegalArgumentException("shmem: filename required for mmap backend");
        }
        break;
      case Config.BACKEND_POSIX_SHM:
        if (cfg.shmName() == null || cfg.shmName().isEmpty()) {
          throw new IllegalArgumentException("shmem: shmName required for posix_shm backend");
        }
        break;
      default:
        throw new IllegalArgumentException("shmem: unknown backend: " + backend);
    }
  }

  private static com.jgshmem.config.Config.Ring toUpstreamRing(Config cfg) {
    com.jgshmem.config.Config.Ring ring = new com.jgshmem.config.Config.Ring();
    ring.backend = cfg.backend() == null ? Config.BACKEND_MMAP : cfg.backend();
    ring.filename = cfg.filename();
    ring.shmName = cfg.shmName();
    ring.capacity = cfg.capacity();
    ring.maxObjectSize = cfg.maxObjectSize();
    return ring;
  }
}
